csrc/optimization/alias_analysis.cpp

```diff
@@ -103,27 +42,151 @@ class AliasFinder : public OptOutConstDispatch {
-    // `in`. Both `in` and `out` are allocated contiguously per the
-    // rfactor domain. Also, the ViewOp can't transform any expanded broadcast
-    // IterDomain.
-    analysis_.add(
-        out,
-        in,
-        {out_allocation,
-         TensorDomain::getContiguityFilledWith(out_allocation, true)});
+  if (!ir_utils::computePermutation(in_rfactor, in_layout.allocation_domain)
+           .has_value()) {
+    // Give up when `in`'s allocation domain is not an rfactor permutation.
+    return;
   }
+
+  std::unordered_map<IterDomain*, IterDomain*> in_rfactor_to_out_root =
+      PairwiseRootDomainMap(in, out).mapBroadcast(true).mapProducerToConsumer();
+
+  // Collect the allocation order of `in`'s rfactor domain and thus `out`'s root
+  // domain.
+  LinkedHashMap<IterDomain*, std::optional<bool>> allocation_to_contiguity;
+  for (const auto i : c10::irange(in_layout.allocation_domain.size())) {
+    IterDomain* in_allocation_id = in_layout.allocation_domain[i];
+    if (!in_rfactor_to_out_root.count(in_allocation_id)) {
+      // `in_allocation_id` is a reduction product.
+      continue;
+    }
+    IterDomain* out_root_id = in_rfactor_to_out_root.at(in_allocation_id);
+    allocation_to_contiguity.pushBack(out_root_id, in_layout.contiguity[i]);
+  }
+
+  // Replay `Expr`s from `out`'s root to `out`'s rfactor on `out`'s root.
+  // Stop when an `Expr` requires a data copy; otherwise generate the allocation
+  // order of `out`'s rfactor domain and the corresponding contiguity flags.
+  for (Expr* transform : DependencyCheck::getAllExprsBetween(
+           {out_root.begin(), out_root.end()},
+           {out_rfactor.begin(), out_rfactor.end()})) {
+    if (Split* split = dynamic_cast<Split*>(transform)) {
+      const auto [contiguity, split_i] =
+          allocation_to_contiguity.erase(split->in());
+      auto [outer_contiguity, inner_contiguity] = splitContiguity(contiguity);
+      allocation_to_contiguity.insert(
+          split_i, split->outer(), outer_contiguity);
+      allocation_to_contiguity.insert(
+          split_i, split->inner(), inner_contiguity);
+    } else if (Merge* merge = dynamic_cast<Merge*>(transform)) {
+      const auto [outer_contiguity, inner_i] =
+          allocation_to_contiguity.erase(merge->outer());
+      if (inner_i == allocation_to_contiguity.end() ||
+          inner_i->first != merge->inner()) {
+        // Outer and inner are not adjacent in allocation order.
+        return;
+      }
+      const auto [inner_contiguity, merge_i] =
+          allocation_to_contiguity.erase(merge->inner());
+      const auto [mergeable, contiguity] = mergeContiguity(
+          merge->outer(), outer_contiguity, merge->inner(), inner_contiguity);
+      if (!mergeable) {
+        return;
+      }
+      allocation_to_contiguity.insert(merge_i, merge->out(), contiguity);
+    } else {
+      NVF_ERROR(
+          false, "Expect Split or Merge, but found: ", transform->toString());
+    }
```

Comment:
The other transform you might expect to find here is a Resize. When there is a resize we are either padding or truncating an axis. I think padding will always require materialization. Truncation followed by a merge with something other than broadcast will require materialization also. If you truncate then split that is fine, but if those split IDs later merge with IDs from another ID then we need to materialize I think.

Reply:
Note that in general we can't tell the sign of the pad values in a Resize. If we can prove both pads are non-positive and there is no downstream merge that has a dependency other than the Resize input we don't have to materialize. This would be the case for a simple slice.

Reply:
Nevermind! I forgot this is a `ViewOp` so it's correct as is. No need to look for `Resize` here. I'll look at #1281 for that.